### ISSUE TITLE
Lost $ 11,926.44 USD caused by bid caching

### ISSUE THREAD

**Issue by freegames66** [prebid/Prebid.js#3342]
I have faced a big problem with bid caching when working with **districTM**. I hope that there is representative of both **districTM** and **Prebid.org** here to help me solve this problem.

Recently, We have been deducted a sum of **$ 11,926.44 USD** by one of our header bidding partner. 

The reason they came up is that we are using bid caching and districTM does not allow that.

### There are some things that I want to clarify:

1 / We did not know about bid caching until my account manager at districTM said that there was abnormal activity on my account and they asked me if I was using bid caching?

2 / Immediately after receiving the email from the account manager, I have been looking for information on bid caching and found out that this is a default function of prebid.js since version 1.0, and currently, there is no setting to disable it. I would like to emphasize that we do not intend to use bid caching with any of our partners.

3 / I have used this prebid setting for a very long time and there is no problems with our partners even including districTM. We did not make any setup changes, however, we have updated to the latest Prebid.js version since we received notice that the prebid.js versions that are prior to version 1.0 will no longer be supported.

### Some of my questions:

1 / My account manager at districTM has told me in the email that: 

> "bids we sent on the 29th of October are winning auctions on October 30th"

Is there anyone can help me explain this?

- In theory, it is impossible. Because bid caching only can be used for a certain amount of time (TTL), and it will  be lost when the user reloads the page. Let's see the document of prebid.org below:

> Does Prebid.js cache bids?
> Yes. As of version 1.0, Prebid.js will re-consider previous bids under limited circumstances. It will cache and reconsider bids in refresh scenarios when the bid is:
> 
> for the same AdUnit,
> on the same page view,
> for the same user, and
> up to a certain Time-to-Live (TTL) or until the bid wins and is displayed.
> Since the storage is in the browser, cached bids only apply to a single page context. **_If the user refreshes the page, the bid is lost._**

2 / According to what I understand, the Prebid.js library only sends bidder information about the user, Prebid.js will receive the responses about the bid from bidders, and then they forward these responses to AdServer. So, how can the Prebid.js library use bid caching to influence to the auction between SSP (here is DistricTM) and DSPs? DistricTM is saying that I have used Bidcaching to change the price that is sent to advertisers (DSPs).

### Setting on my header bidding wrapper:
1/ This is my old prebid config that I used and it consists of the feature auto refresh adUnits after every 120 seconds.

https://jsfiddle.net/freegames66/dnpgLbfa/

2/ This is my new prebid config and the solution to disable the bid caching

https://jsfiddle.net/freegames66/2pouv81c/

I have created an array consists of bidders that allow to refresh: 
`var refreshAllowed = ['rhythmone','conversant'];`

In the ad refresh function, I remove bidders that don't allow to refresh in adUnits and recreate `pbjs.adUnits = [];` after that, recall `pbjs.addAdUnits(adUnits);` before calling `pbjs.requestBids`
            
- `Use pbjs.getBidResponsesForAdUnitCode(currentDivId);` instead of `pbjs.getHighestCpmBids(currentDivId);`
- Do not use `pbjs.renderAd`, instead of that, let's use:
```
iframeDoc.write(bidWon.ad); 
iframeDoc.close();
```

### Final words
I take this issue up here for us to look at and find out the real cause of this problem together.

My point of view is that if this is really a problem that caused by my prebid settings or this is an error on prebid.js, I totally agree with the deduction that is given by districTM. 

However, if this is a bug caused by districTM's system, I would like to get the fairness. 

I have been working with the Prebid.js library for a long time and understand a part of the way that it actives. Header Bidding in general and Prebid.js in particular have opened up a new revolution. However, still there are lots of inadequacies and bugs that we need to contribute together and build it more complete.

**Comment by chadex** [comment]
You want to verify the volume of cases through analysis of the the log-level data. It's important you agree on the specific cases warranting a each billing adjustment. I've launched a log-level reporting service Viewex Bid Data specifically for this purpose. I'm happy to look over the data FOC. Email me on lee@viewex.co.uk

**Comment by rale** [comment]
It was probably due to the changes that happened with getHighestCpmBids. Take a look at #2959.

**Comment by df-districtm** [comment]
I feel I need to clarify some things here, as CIO of District M.

First, your assertion that you

>  Lost $ 11,926.44 USD caused by bid caching

 is wrong.

Second, this was for another account under the name BidGear, which we just now realize was the same person as FreeGames666, another site where we noticed the same behaviour to a lesser extent. I feel this needs to be clarified.

The situation we identified was the following, which we attributed to some sort of bid caching, or other method of triggering impressions beyond their Time To Live. The real mechanism of which we haven't investigated further, our main focus was to remove this source of supply which we considered to be fraudulent.

Sometime in October, we started noticing higher discrepancies than usual with our DSP partners. After investigating, this was mainly coming from one publisher, BidGear. The way the impressions were measured was through ad markup, with a notification going to our DSP partner and another one going to our notification endpoint upon rendering the ad.

Upon investigating further, a DSP partner provided insight about a higher than usual number of notifications coming after TTL (of 3 minutes). This prompted us to verify this and implement a new notification endpoint that would enforce the same TTLs as our DSP partners to curb discrepancies as well as stronger validation to prevent such behavior.

When this new endpoint was released, under a distinct URL, we kept receiving notifications from BidGear (and FreeGames666 in a marginal way) on the previous the endpoint. This went on **FOR DAYS**, which prompted further analysis and upon review, our team determined there was a difference of 11,926.44$ between what was received on our previous version VS what our DSP partners were recording. 

So, when you say 

>  So, how can the Prebid.js library use bid caching to influence to the auction between SSP (here is DistricTM) and DSPs? DistricTM is saying that I have used Bidcaching to change the price that is sent to advertisers (DSPs).

This is not what our team claimed. They never claimed there was interference in the auction mechanism, rather than rendering the ads hours and days after the auction occurred, caused a wrongful attribution or revenue, which advertisers never were charged for, therefore, we had to adjust your earnings accordingly.

Long story short, this is a typical invalid traffic clawback, not a software bug. We have made this clear in our communications with you. If necessary, we can disclose them for transparency, but we would appreciate that this issue gets closed as it is detrimental to our company's reputation.

**Comment by bretg** [comment]
@freegames66 - While Prebid.org can't mediate this issue between you and the SSP/DSP, we do appreciate the chance to make sure there aren't issues in Prebid.js.

@df-districtm has clarified the source of the discrepancy as "notifications for bids generated back when it was providing the initial notification pixel"... days earlier. So this shouldn't be related to the Prebid.js bid reuse feature, which is constrained to refreshes and within a timespan of minutes, not days.

Upon further review, we found several bidders that define their TTLs to be very long, which was not caught during adapter review. This could be related to a documentation issue that was closed yesterday (https://github.com/prebid/prebid.github.io/pull/1047) where one of the places that lists TTL showed it as milliseconds rather than seconds. The main "how to build an adapter" page was correct, but the "Publisher API" page had the wrong unit. But DistrictMDMX's adapter has a TTL of 60 seconds, which rules out adapter TTL as a cause in this specific case. Still, as a result of this investigation, we will contact each of the adapters having over-long TTLs.

I'm sure there's a valid technical explanation for this issue. e.g... I've seen whole pages accidentally copied onto a CDN, ads and all. Or perhaps there was a proxy issue for a subset of your users. Our suggested next step for you is to work with your partners to get log-level data and see if there's a pattern of source IP address, URL, or ad ID.

**Comment by freegames66** [comment]
> I feel I need to clarify some things here, as CIO of District M.
> 
> First, your assertion that you
> 
> > Lost $ 11,926.44 USD caused by bid caching
> 
> is wrong.
> 
> Second, this was for another account under the name BidGear, which we just now realize was the same person as FreeGames666, another site where we noticed the same behaviour to a lesser extent. I feel this needs to be clarified.
> 
> The situation we identified was the following, which we attributed to some sort of bid caching, or other method of triggering impressions beyond their Time To Live. The real mechanism of which we haven't investigated further, our main focus was to remove this source of supply which we considered to be fraudulent.
> 
> Sometime in October, we started noticing higher discrepancies than usual with our DSP partners. After investigating, this was mainly coming from one publisher, BidGear. The way the impressions were measured was through ad markup, with a notification going to our DSP partner and another one going to our notification endpoint upon rendering the ad.
> 
> Upon investigating further, a DSP partner provided insight about a higher than usual number of notifications coming after TTL (of 3 minutes). This prompted us to verify this and implement a new notification endpoint that would enforce the same TTLs as our DSP partners to curb discrepancies as well as stronger validation to prevent such behavior.
> 
> When this new endpoint was released, under a distinct URL, we kept receiving notifications from BidGear (and FreeGames666 in a marginal way) on the previous the endpoint. This means we were receiving notifications that for bids generated **back when it was providing the initial notification pixel**. This went on **FOR DAYS**, which prompted further analysis and upon review, our team determined there was a difference of 11,926.44$ between what was received on our previous version VS what our DSP partners were recording.
> 
> So, when you say
> 
> > So, how can the Prebid.js library use bid caching to influence to the auction between SSP (here is DistricTM) and DSPs? DistricTM is saying that I have used Bidcaching to change the price that is sent to advertisers (DSPs).
> 
> This is not what our team claimed. They never claimed there was interference in the auction mechanism, rather than rendering the ads hours and days after the auction occurred, caused a wrongful attribution or revenue, which advertisers never were charged for, therefore, we had to adjust your earnings accordingly.
> 
> Long story short, this is a typical invalid traffic clawback, not a software bug. We have made this clear in our communications with you. If necessary, we can disclose them for transparency, but we would appreciate that this issue gets closed as it is detrimental to our company's reputation.

Thank you for your answer. However, I want to clarify some of the issues:

1 / Bidgear and Freegames66 are two projects owned by X8media JSC. I am responsible for the project Freegames66. Because these two projects are separate, so we have created two separate districtm accounts for reporting revenue for each project. 

2 / I did not receive any email about invalid traffic issue. My account manager only told me about the bidcaching issue. So, I got up here to ask for help from the prebid community.


3 / About the issue of invalid traffic on Bidgear account:
- We have been working with you for a long time, and our inventory volume, as well as the list of sites that are working with you, are almost unchanged.
- We have maintained a prebid setup for a very long time and have not had any problems with invalid traffic until recently.

4 / About the issue of invalid traffic on Freegames66 account:
- As you know, most of our websites on this account are in the field of gaming.
We can offer you full statistics on Google analytics and you will see that > 80% of our traffic comes from search engines. The rest one comes from direct and a very small amount of traffics comes from the referral source. I am the direct manager of Freegames66's projects, so I can guarantee that 100% of our traffic is organic.

We also work with lots of other partners such as AOL, Conversantmedia, GumGum, Rhythmone, EMX digital, and do not have any problems with them.
- I have just received an email from Account Manager (Freegames66) and he tells me that the revenue that is deducted in November is about 75%. This is the same meaning as about  75% of my traffic is invalid traffic. And I feel it is not convincing at all because, as I said before, 80% of our traffic comes from the search engine (Google). And one thing that makes me wonder is that why this issue only happens to DistricTM without happening with my other partners?

5 / We do not intentionally do anything to increase the revenue invalidly.
